text/0000-kafka-topics-per-tenant.md

```diff
@@ -0,0 +1,121 @@
+On the client side, buffers are created for each partition in memory. So a producer that interacts with very many thousands of partitions will consume more memory. To subscribe to a number of partitions with a regex subscription pattern, Kafka will send the total list of topics to each consumer. A consumer would then filter the list to determine the topics that match the subscription pattern. Sending a total list of topics would require some consensus from all brokers, therefore overhead. Consumers would ask for the full list at regular interval to catch newly added topics.
+- **Cost**: Kafka's use in FOLIO is still limited, more FOLIO modules will continue to employ event driven techniques that will cause an explosion of partitions. For hosting providers that self host Kafka, more brokers or broker capacity will be needed to service load generated by the high number of partitions. For hosting providers who utilize cloud managed Kafka, cloud providers typically charge by partition or set limits on partition count for a particular broker size. For example, Confluent Cloud charges $13/partition/year with a maximum of 4096 partitions for one cluster. AWS charges $3679/year for a broker instance with 4 vCPU and 16GB Memory and a soft limit of 1000 partitions per broker. This does not include extra brokers in other availability zones to support high availability.
+
+## Out of Scope
+- Enabling tenant seperation within a module. Tenant seperation meaning "How do we prevent a tenant's data from being processed in place of another?", "How can we prevent a module(specific to a tenant) from seeing data from other tenants?"
+- Moving forward with the [Temporary Kafka Security Solution](https://wiki.folio.org/x/YYVFAw). More details are included in the *Related Concerns* section of this document.
+- Countering module multi-versioning i.e. more than one version of a module is installed in FOLIO with tenants being able to target specific versions of a module. **Module multi-versioning is supported by OKAPI & mod-pubsub but not for modules that interact directly via Kafka. This is an existing deficiency prior to the changes documented by this RFC.** More details are included in the *Related Concerns* section of this document.
+
+## Detailed Explanation/Design
+To promote flexibility of FOLIO implementation patterns, the switch to tenant collection topics will be configurable via environment variables. The current naming scheme for topics is like so:
+>[environment].[namespace].[tenant].[eventtype]
+
+An example: `folio.Default.diku.DI_COMPLETED`. The proposal is to preserve the naming structure by declaring one more "tenant" where messages for all tenants on the cluster is received for an application concern. e.g. `folio.Default.ALL.DI_COMPLETED`.<p>
```

Comment:
I have made changes.

I will admit that I was not expecting varying values for the collection name being used by hosting providers. I will also admit that it can give rise to some sophisticated configuration options like Fast lane/prioritized topics. I believe more work would still need to be done to achieve these lofty goals.